## 1. Cities in Sweden - create dataset (*)

Create this DataFrame from scratch:

![](../assets/print1.PNG)

a) Use your DataFrame to print out all the cities. (*)

  b) Select only the row which contains Göteborg. Do this by using the name Göteborg. (*)

  c) Sort the cities by population from largest to smallest. (*)

  d) Filter out the three largest cities. (*)

  e) The whole population in Sweden 2020 is 10379295. Use this number to create a new column in your sorted DataFrame named: Population (%). This column should be filled with percentage of the Swedish population for each city.

In [101]:
# Create DataFrame from scratch:

import pandas as pd  # Imports Pandas


kommun = pd.Series({0: "Malmö", 1: "Stockholm", 2: "Uppsala", 3: "Göteborg"})
population = pd.Series({0: 347949, 1: 975551, 2: 233839, 3: 583056})


df_städer = pd.DataFrame(
    {"Kommun": kommun, "Population": population}
)  # Creates a Dataframe from my 2 variables, kommun, populaion
df_städer

Kommun  Population
0      Malmö      347949
1  Stockholm      975551
2    Uppsala      233839
3   Göteborg      583056

a) Use your DataFrame to print out all the cities

In [102]:
# a) Use your DataFrame to print out all the cities.

df_städer["Kommun"].unique()   # Print out all cities

array(['Malmö', 'Stockholm', 'Uppsala', 'Göteborg'], dtype=object)

In [112]:
print(df_städer.loc[0:0])  # Finds row and columns

  Kommun  Population
0  Malmö      347949


In [104]:
df_städer.loc[[3]] #Find row 3 with both columns(Kommun, Population)

Kommun  Population
3  Göteborg      583056

b) Select only the row which contains Göteborg. Do this by using the name Göteborg.

In [113]:
df_städer["Kommun"] #Just printing out to see dtype and get the list of the cities

0        Malmö
1    Stockholm
2      Uppsala
3     Göteborg
Name: Kommun, dtype: object

In [115]:
# Selects row where "Kommun" column is equal to Göteborg

df_städer.loc[df_städer['Kommun'] == "Göteborg"]

Kommun  Population
3  Göteborg      583056

In [120]:
## Testar hur iloc fungerar ##

# Selects columns in position 0
df_städer.iloc[:, [0]]  # Denna raden gör att jag kan hitta t ex kommun bara, eller bara population och få siffrorna. 
# Testa skriva 1 istället för 0 om något är oklart på vad denna gör.

Kommun
0      Malmö
1  Stockholm
2    Uppsala
3   Göteborg

c) Sort the cities by population from largest to smallest

In [122]:
df_städer.info() # Gets some info

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4 entries, 0 to 3
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Kommun      4 non-null      object
 1   Population  4 non-null      int64 
dtypes: int64(1), object(1)
memory usage: 268.0+ bytes


In [132]:
df_städer_sorted = df_städer.sort_values(by="Population", ascending = False) # Sorts values by column Population (largest first, could reverse it by making it True)
df_städer_sorted


Kommun  Population
1  Stockholm      975551
3   Göteborg      583056
0      Malmö      347949
2    Uppsala      233839

# d) Filter out the three largest cities. 

In [161]:
df_städer_sorted.tail(-3)  # Returns row with tail from position (-3)

Kommun  Population
2  Uppsala      233839

In [162]:
df_städer_sorted.iloc[:3]  # Finds the three largest based on iloc position finder

Kommun  Population
1  Stockholm      975551
3   Göteborg      583056
0      Malmö      347949

In [163]:
df_städer_sorted.head(3)   # Using head to get the first biggest cities in the städer sorted

Kommun  Population
1  Stockholm      975551
3   Göteborg      583056
0      Malmö      347949

In [149]:
df_städer_sorted["Population"].value_counts() # Counts the how many time the unique values are counted

975551    1
583056    1
347949    1
233839    1
Name: Population, dtype: int64

In [153]:
(df_städer_sorted["Population"] >= 300000).value_counts()  #Sorts out by value. As i set the bar so high i know that the smallest city of the 4 will be false. 

True     3
False    1
Name: Population, dtype: int64

In [167]:
df_städer.info

<bound method DataFrame.info of       Kommun  Population
0      Malmö      347949
1  Stockholm      975551
2    Uppsala      233839
3   Göteborg      583056>

e) The whole population in Sweden 2020 is 10379295. Use this number to create a new column in your sorted DataFrame named: Population (%). This column should be filled with percentage of the Swedish population for each city.

In [183]:
# e) The whole population in Sweden 2020 is 10379295.
# Use this number to create a new column in your sorted DataFrame named: Population (%). This column should be filled with percentage of the Swedish population for each city.
asd = 10379295

df_städer2 = df_städer_sorted.assign(Population_in_percent=lambda x: population / asd * 100)   # I have trouble renaming my due to % disrupt everyting?
df_städer2

Kommun  Population  Population (%)  Population_in_percent
1  Stockholm      975551        9.399010               9.399010
3   Göteborg      583056        5.617491               5.617491
0      Malmö      347949        3.352338               3.352338
2    Uppsala      233839        2.252937               2.252937

In [184]:
asd2 = 10379295

df_städer_sorted["Population (%)"] = (df_städer_sorted["Population"] / asd2 * 100)
df_städer_sorted

Kommun  Population  Population (%)
1  Stockholm      975551        9.399010
3   Göteborg      583056        5.617491
0      Malmö      347949        3.352338
2    Uppsala      233839        2.252937

In [ ]:
# Just for fun to print
print(f"{städer_series[0] = }")
print(f"{städer_series[1] = }")
print(f"{städer_series[2] = }")
print(f"{städer_series[3] = }")

print(f"{städer_series.keys()[1] = }")  # Plockar ut ett enskilt värde

städer_series[0] = 347949
städer_series[1] = 97551
städer_series[2] = 233839
städer_series[3] = 583056
städer_series.keys()[1] = 'Stockholm'


---

## 2. Cities in Sweden - real dataset (*)

Download the file komtopp50_2020.xlsx from the course github repo. The data is taken from SCB. 

&nbsp; a) Read in the tab "Totalt" into a DataFrame and start exploring the data with some simple explorations such as

```python
df.head()
df.info()
df.describe()
```

Feel free to do more explorations. (*)

&nbsp; b) Clean your data so that the head looks like this: (*)


|    |   Rang 2020 |   Rang 2019 | Kommun   |   Folkmängd 2020 |   Folkmängd 2019 |   Förändring |
|---:|------------:|------------:|:---------|-----------------:|-----------------:|-------------:|
|  0 |          83 |          84 | Ale      |            31868 |            31402 |     1.48398  |
|  1 |          64 |          64 | Alingsås |            41602 |            41420 |     0.439401 |
|  2 |         123 |         123 | Alvesta  |            20224 |            20134 |     0.447005 |
|  3 |         255 |         255 | Aneby    |             6821 |             6848 |    -0.394276 |
|  4 |         169 |         167 | Arboga   |            14039 |            14087 |    -0.34074  |

&nbsp; c) Sort the cities by population from largest to smallest. (*)

&nbsp; d) Filter out the five smallest cities. (*)

&nbsp; e) Use the DataFrame to calculate Sweden's population in 2019 and 2020. (*)

&nbsp; f) Plot a bar chart for the five largest cities and the five smallest cities. (*) 


In [12]:
# a) read in data file.
import pandas as pd

# TODO : Read in "Totalt" into a df. Do df.head(), df.info(). df.describe()
df = pd.read_excel("../Data/komtopp50_2020.xlsx", sheet_name="Totalt", header=[5,6])  # Reading in tab "Totalt"  # Header reads in from row 5, fattar inte riktigt varför just 6. Men its working i guess

df.head(10)

# We can immediatyly see we have a lot of missing values.
# We have columns that are unnamed 
# TODO: Columns: "Rang2020" -> Rang2019 -> Kommun -> folkmängd 2020 -> Folkmängd 2019 -> Förändring
# TODO: Remove row 0-4
# We have the rows

Rang                  Kommun Folkmängd 31 december        Förändring
  2020 2019 Unnamed: 2_level_1                  2020   2019          %
0   83   84                Ale                 31868  31402   1.483982
1   64   64           Alingsås                 41602  41420   0.439401
2  123  123            Alvesta                 20224  20134   0.447005
3  255  255              Aneby                  6821   6848  -0.394276
4  169  167             Arboga                 14039  14087  -0.340740
5  287  287           Arjeplog                  2718   2785  -2.405745
6  261  261         Arvidsjaur                  6145   6220  -1.205788
7  103  102             Arvika                 25932  26045  -0.433864
8  198  199          Askersund                 11471  11377   0.826228
9  112  112             Avesta                 23067  23178  -0.478902

# b)  your data so it looks like(see above):

In [20]:
#  I want to delete some labels in the columns.

df2 = df[:]  # Returns all elements -> [:]
df2

names = [
    "Rang 2020",
    "Rang 2019",
    "Kommun",
    "Folkmängd 2020",
    "Folkmängd 2019",
    "Förändring %",
]

df2.columns = names  # Gives list into the columns. 
df2.head(10)

Rang 2020  Rang 2019      Kommun  Folkmängd 2020  Folkmängd 2019  \
0         83         84         Ale           31868           31402   
1         64         64    Alingsås           41602           41420   
2        123        123     Alvesta           20224           20134   
3        255        255       Aneby            6821            6848   
4        169        167      Arboga           14039           14087   
5        287        287    Arjeplog            2718            2785   
6        261        261  Arvidsjaur            6145            6220   
7        103        102      Arvika           25932           26045   
8        198        199   Askersund           11471           11377   
9        112        112      Avesta           23067           23178   

   Förändring %  
0      1.483982  
1      0.439401  
2      0.447005  
3     -0.394276  
4     -0.340740  
5     -2.405745  
6     -1.205788  
7     -0.433864  
8      0.826228  
9     -0.478902

&nbsp; c) Sort the cities by population from largest to smallest. (*)

In [51]:
df2_sorted = df2.sort_values(by="Rang 2020", ascending = True, ignore_index=True)   # Same as i did in the first assignment, but added ignore_index = True

# ignore_index gör så att du kan kontrollera det nya registret(indexet) av den nya outputen. Default är false. Eftersom jag satte True så kommer pandas ignorera index värdena i input och
# ge ett nytt index för outputen (Den stapeln längst till vänster som är 0, 1, 2 osv.)

df2_sorted.head()


Rang 2020  Rang 2019     Kommun  Folkmängd 2020  Folkmängd 2019  \
0          1          1  Stockholm          975551          974073   
1          2          2   Göteborg          583056          579281   
2          3          3      Malmö          347949          344166   
3          4          4    Uppsala          233839          230767   
4          5          5  Linköping          164616          163051   

   Förändring %  
0      0.151734  
1      0.651670  
2      1.099179  
3      1.331213  
4      0.959822

&nbsp; d) Filter out the five smallest cities. (*)

In [55]:
df3_sorted = df2.sort_values(by="Rang 2020", ascending = False, ignore_index=True)
df3_sorted.head(5)

Rang 2020  Rang 2019    Kommun  Folkmängd 2020  Folkmängd 2019  \
0        290        290  Bjurholm            2387            2408   
1        289        289   Sorsele            2442            2489   
2        288        288   Dorotea            2498            2551   
3        287        287  Arjeplog            2718            2785   
4        286        286     Åsele            2805            2794   

   Förändring %  
0     -0.872093  
1     -1.888309  
2     -2.077617  
3     -2.405745  
4      0.393701

&nbsp; e) Use the DataFrame to calculate Sweden's population in 2019 and 2020. (*)

In [56]:
df2_sorted.sum()

Rang 2020                                                     42195
Rang 2019                                                     42194
Kommun            StockholmGöteborgMalmöUppsalaLinköpingÖrebroVä...
Folkmängd 2020                                             10379295
Folkmängd 2019                                             10327589
Förändring %                                              44.854988
dtype: object

&nbsp; f) Plot a bar chart for the five largest cities and the five smallest cities. (*) 

In [ ]:
# YEESSS SSIR TIME TO PLOT THIS SHIET

import matplotlib.pyplot as plt

fig, axes = plt.subplots(1,3, figsize = (16), )

## 3. Cities in Sweden - gender (*)

We continue with the same Excel-file as in task 2, but now you should also read in the sheets "Kvinnor" and "Män" into two additional DataFrames. In this task, many operations are similar to all three datasets, try creating custom made functions to reuse as much code as possible.

&nbsp; a) Clean your data so that the head looks like this: (*)

Male: 

|    |   Rang 2020 |   Rang 2019 | Kommun   |   Folkmängd 2020 |   Folkmängd 2019 |   Förändring | Kön   |
|---:|------------:|------------:|:---------|-----------------:|-----------------:|-------------:|:------|
|  0 |          81 |          83 | Ale      |            16256 |            16054 |     1.25825  | Man   |
|  1 |          64 |          64 | Alingsås |            20702 |            20646 |     0.271239 | Man   |
|  2 |         122 |         123 | Alvesta  |            10367 |            10339 |     0.270819 | Man   |
|  3 |         255 |         255 | Aneby    |             3503 |             3498 |     0.142939 | Man   |
|  4 |         169 |         168 | Arboga   |             7094 |             7114 |    -0.281136 | Man   |

Female: 

|    |   Rang 2020 |   Rang 2019 | Kommun   |   Folkmängd 2020 |   Folkmängd 2019 |   Förändring | Kön    |
|---:|------------:|------------:|:---------|-----------------:|-----------------:|-------------:|:-------|
|  0 |          84 |          85 | Ale      |            15612 |            15348 |     1.72009  | Kvinna |
|  1 |          64 |          64 | Alingsås |            20900 |            20774 |     0.606527 | Kvinna |
|  2 |         123 |         123 | Alvesta  |             9857 |             9795 |     0.632976 | Kvinna |
|  3 |         255 |         255 | Aneby    |             3318 |             3350 |    -0.955224 | Kvinna |
|  4 |         165 |         164 | Arboga   |             6945 |             6973 |    -0.401549 | Kvinna |


&nbsp; b) Merge the male and female DataFrames vertically and set index to "Kommun". Note that there now should be 580 rows now.  (*)

|    | Kommun   |   Folkmängd 2020 |   Folkmängd 2019 |   Förändring | Kön   |
|---:|:---------|-----------------:|-----------------:|-------------:|:------|
|  0 | Ale      |            16256 |            16054 |     1.25825  | Man   |
|  1 | Alingsås |            20702 |            20646 |     0.271239 | Man   |
|   ...  | ...       | ...  | ...   | ...  | ...    |
| 288 | Överkalix    |             1559 |             1578 |    -1.20406  | Kvinna |
| 289 | Övertorneå   |             2027 |             2065 |    -1.84019  | Kvinna |

&nbsp; c) Extract and change column name from the total DataFrame so that the head look like this: (*) 

|    | Kommun   |   Total Pop 2020 |   Total Pop 2019 |   Total förändring |
|---:|:---------|-----------------:|-----------------:|-------------------:|
|  0 | Ale      |            31868 |            31402 |           1.48398  |
|  1 | Alingsås |            41602 |            41420 |           0.439401 |
|  2 | Alvesta  |            20224 |            20134 |           0.447005 |
|  3 | Aneby    |             6821 |             6848 |          -0.394276 |
|  4 | Arboga   |            14039 |            14087 |          -0.34074  |

&nbsp; d) Merge this data with the data in b) so that the head look like this: (*)

| Kommun    |   Folkmängd 2020 |   Folkmängd 2019 |   Förändring | Kön    |   Total Pop 2020 |   Total Pop 2019 |   Total förändring |
|:----------|-----------------:|-----------------:|-------------:|:-------|-----------------:|-----------------:|-------------------:|
| Stockholm |           482982 |           482220 |     0.158019 | Man    |           975551 |           974073 |           0.151734 |
| Stockholm |           492569 |           491853 |     0.145572 | Kvinna |           975551 |           974073 |           0.151734 |
| Göteborg  |           292505 |           290308 |     0.756782 | Man    |           583056 |           579281 |           0.65167  |
| Göteborg  |           290551 |           288973 |     0.546072 | Kvinna |           583056 |           579281 |           0.65167  |
| Malmö     |           175411 |           173543 |     1.07639  | Kvinna |           347949 |           344166 |           1.09918  |

&nbsp; e) Create barplots showing the gender populations of Swedens 10 largest and 10 smallest cities. (*) 
 
&nbsp; f) Create a pie chart showing the total male and female population in Sweden 2020. (*)

&nbsp; g) Create a barplot showing the cities with the five largest percentual gender difference in 2020. (**)

&nbsp; h) Create a barplot showing the top 5 cities with largest populational growth from 2019 to 2020 (**)

&nbsp; i) Feel free to investigate other questions you are interested in using these datasets. (*), (**)

<br/>